### PR TITLE
Added vignette explaining custom variables

### PR DESCRIPTION
**Pull request by yulric**
This PR goes over a new custom variable feature to deal with this [issue](https://github.com/Big-Life-Lab/recodeflow/issues/22).

It has an .Rmd file and a set .CSV files referenced by it.

### CODE REVIEW

**Review by Rhan43 — CHANGES_REQUESTED** [review]
Thanks, Yulric! The logic makes sense, and I think it's easy to follow. I only have a few comments.

**Comment by yulric** [comment]
Thanks @Rhan43, I've incorporated your comments. I'll get @DougManuel and @CBjerke to take a look before merging this in and implementing it into recodeflow.

**Comment by Rhan43** [comment]
@yulric, does the variable sheet look the same as before, or do we need to add a "customVariable" column in the variables sheet too? (the vignette only references the variable_detials sheet)

**Comment by yulric** [comment]
> @yulric, does the variable sheet look the same as before, or do we need to add a "customVariable" column in the variables sheet too? (the vignette only references the variable_detials sheet)

Since the custom variable does not actually exist in the database, I don't think it needs to be in the variables sheet. Only the final variable, `primary_lang` and `secondary_lang` should be in there. What do you think?

**Comment by Rhan43** [comment]
> > @yulric, does the variable sheet look the same as before, or do we need to add a "customVariable" column in the variables sheet too? (the vignette only references the variable_detials sheet)
> 
> Since the custom variable does not actually exist in the database, I don't think it needs to be in the variables sheet. Only the final variable, `primary_lang` and `secondary_lang` should be in there. What do you think?

Makes sense to me! Okay, I'll only add the "customVariable" to the variable_details

**Review by DougManuel — CHANGES_REQUESTED** [review]
Comments on function:

1) This feature shares categories between variables. There are quite a few similar examples, with even longer lists of categories. Many of the category lists are well-established standards for coding health. Hospital discharge databases have multiple diagnosis fields that use ICD9 or 10, each with over 1000+ categories; there are similar therapeutic variables; LOINC for lab values, drugs with thousands of DIN or ATC codes, etc. 

What about having a new sheet called reference categories — or a similar name? There would be fewer headers than variable details (which have many headers needed for the language categories) but potentially a few additional headers like an external reference link. 

2) The new “customVariable” header is a good idea, but I wonder about making it potentially used for more than this single-use case of shared categories. I have a potentially new feature that we will consider adding that may require additional support. Call this header ‘rules’ or ‘templates’. Instead of customVariable = “language”, have rule = catRefList::language. 

With suggestion 1, varible_details becomes cleaner: the long category lists are removed, and you don’t require fields with responses like “yes” and “no”. 

Comments on documentation.

1) Can you find a name that is more specific than ‘custom variable’? The term custom variable has many potential meanings. What about “template variable”, “shared categories,” …

2) The first sentence also has several non-specific terms that make it hard to follow.  I’ve made suggestions for the first paragraph.

"Shared Categories," simplifies data transformation when multiple variables use the same set of standard categories. The Shared Categories feature allows users to establish a single, reusable set of recoding rules for standard categories, eliminating redundancy and reducing the potential for error. This is especially useful when working with large, complex datasets that require consistent recoding across multiple variables.
The use case described below is when a respondent has two language variables (primary and secondary language spoken) which share the same list of languages. Other examples include country codes, product categories, drug or medication lists. 
3) The document could be organized into sections, i.e. Introduction, Workflow, Configuration, Examples. 
4) It may help to have a visual aid like a diagram or flowchart.

File: vignettes/custom-variables.Rmd [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+The `customVariable` column can have one of the following values:
+
+1. **Yes**: Value to set when the row is defining a custom variable
+2. **No**: Value to set when the row is defining a normal variable that **does not** extend a custom variable
+3. **<custom_variable_name>**: The name of a custom variable that this row is extending
+
+When creating a custom variable, 
+
+1. The `variable` column should the name of the custom variable
```
Comment (reikookamoto) [inline vignettes/custom-variables.Rmd:71]:
```suggestion
1. The `variable` column should be the name of the custom variable
```

File: vignettes/custom-variables.Rmd [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+2. The `customVariable` column should be Yes
+3. The `variableStart` and `variableStartLabel` column should be `N/A`. Variables that extend the custom variable will set them.
+4. All the other columns should be completed as normal
+
+When extending a custom variable,
+
+1. The custom variable column should be set to the name of the custom variable being extended
+2. The `typeEnd`, `typeStart`, `databaseStart` columns should be equal to value set in the custom variable. The `numValidCat`, `catLabel`, `catLabeLong`, `units`, `recStart`, and `catStartLabel` should set to N/A. These will come from the custom variable.
```
Comment (reikookamoto) [inline vignettes/custom-variables.Rmd:79]:
```suggestion
2. The `typeEnd`, `typeStart`, `databaseStart` columns should be equal to value set in the custom variable. The `numValidCat`, `catLabel`, `catLabeLong`, `units`, `recStart`, and `catStartLabel` should be set to N/A. These will come from the custom variable.
```

**Review by reikookamoto — COMMENTED** [review]
I fixed some typos in the vignette.

While I understand this current implementation reduces the number of rows in a details sheet, I wonder if there's other design choices that can be taken:
- Could we move the information contained in the first 6 rows of [`custom-variable-variable-details.csv`](https://github.com/Big-Life-Lab/recodeflow/blob/639abdef01d4b153e65d5f4917b83b350c8cd859/assets/vignettes/custom-variables/custom-variable-variable-details.csv#L2-L7) to somewhere else so that `variable` column of the details sheet *only* contains variables that appear in the recoded dataset? That's to say, can we store the information about the recycled specification elsewhere?
- Alternatively, we could define a separate function that users can use to automatically populate details sheets:
```
add_repeats(
 details_sheet = "details_sheet.csv", 
 vars = c("primary_language", "second_language"),
 labels = c("Primary Language", "Second Language",
 spec = language
)
```
where `language` is the first 6 rows of [`custom-variable-variable-details.csv`](https://github.com/Big-Life-Lab/recodeflow/blob/639abdef01d4b153e65d5f4917b83b350c8cd859/assets/vignettes/custom-variables/custom-variable-variable-details.csv#L2-L7)

**Comment by reikookamoto** [comment]
Yesterday @DougManuel and I revisited the discussion we had on November 9 about this feature. We think that the best way to proceed is to go with the proposed details sheet design but to give the feature a more descriptive name. 

Alternatives include:

- templateVariable
- templateFactor
- blueprintVariable
- blueprintFactor
- cookieCutter 

Please comment on your preference or provide alternatives by December 22. Beyond this deadline silence is considered a vote on the most popular option.

Once a more descriptive name has been chosen, files in this PR and #24 will be edited to reflect the name change before being merged into master.

**Comment by Rhan43** [comment]
Re: CustomVariable name, my favourite and the one I understand right away is cookieCutter.

**Comment by reikookamoto** [comment]
Forgot to add my own vote earlier for cookieCutter :cookie:

**Comment by yulric** [comment]
I like `templateVariable`

**Comment by DougManuel** [comment]
I like sharedCategories, templateVariable, cookieCutter. Of these three, sharedCategories seems the most explicit. There are lots of different ways to `template`, and cookieCutter does have a nice image and is memorable, but would need an good explaination.

**Comment by CBjerke** [comment]
While 'cookieCutter' has appeal, my vote is for 'templateVariable'. Maintly b/c, as Doug points out, cookieCutter may require more explanation...it is darn cute 'tho!

**Review by CBjerke — APPROVED** [review]
Looks good
